packages/techdocs-common/src/git-auth.ts

```diff
@@ -34,85 +41,67 @@ export function getGitRepoType(url: string): string {
   config: Config,
   host: string,
-): string | undefined {
-  const providerConfigs =
-    config.getOptionalConfigArray('integrations.github') ?? [];
-
-  const hostConfig = providerConfigs.filter(
-    providerConfig => providerConfig.getOptionalString('host') === host,
+): GitHubIntegrationConfig => {
+  const allGitHubConfigs = readGitHubIntegrationConfigs(
+    config.getOptionalConfigArray('integrations.github') ?? [],
   );
-  const token =
-    hostConfig[0]?.getOptionalString('token') ??
-    config.getOptionalString('catalog.processors.github.privateToken') ??
-    config.getOptionalString('catalog.processors.githubApi.privateToken') ??
-    process.env.GITHUB_TOKEN;
-
-  return token;
-}
+  const gitHubIntegrationConfig = allGitHubConfigs.find(v => v.host === host);
+  if (!gitHubIntegrationConfig) {
+    throw new Error(`Unable to locate GitHub integration for the host ${host}`);
+  }
+  return gitHubIntegrationConfig;
+};
 
-export function getGitlabHostToken(
+export const getGitLabIntegrationConfig = (
   config: Config,
   host: string,
-): string | undefined {
-  const providerConfigs =
-    config.getOptionalConfigArray('integrations.gitlab') ?? [];
-
-  const hostConfig = providerConfigs.filter(
-    providerConfig => providerConfig.getOptionalString('host') === host,
+): GitLabIntegrationConfig => {
+  const allGitLabConfigs = readGitLabIntegrationConfigs(
+    config.getOptionalConfigArray('integrations.gitlab') ?? [],
   );
-  const token =
-    hostConfig[0]?.getOptionalString('token') ??
-    config.getOptionalString('catalog.processors.gitlab.privateToken') ??
-    config.getOptionalString('catalog.processors.gitlabApi.privateToken') ??
-    process.env.GITLAB_TOKEN;
-
-  return token;
-}
+  const gitLabIntegrationConfig = allGitLabConfigs.find(v => v.host === host);
+  if (!gitLabIntegrationConfig) {
+    throw new Error(`Unable to locate GitLab integration for the host ${host}`);
+  }
+  return gitLabIntegrationConfig;
+};
 
-export function getAzureHostToken(
+export const getAzureIntegrationConfig = (
   config: Config,
   host: string,
-): string | undefined {
-  const providerConfigs =
-    config.getOptionalConfigArray('integrations.azure') ?? [];
-
-  const hostConfig = providerConfigs.filter(
-    providerConfig => providerConfig.getOptionalString('host') === host,
+): AzureIntegrationConfig => {
+  const allAzureIntegrationConfig = readAzureIntegrationConfigs(
+    config.getOptionalConfigArray('integrations.azure') ?? [],
   );
-  const token =
-    hostConfig[0]?.getOptionalString('token') ??
-    config.getOptionalString('catalog.processors.azureApi.privateToken') ??
-    process.env.AZURE_TOKEN;
-
-  return token;
-}
+  const azureIntegrationConfig = allAzureIntegrationConfig.find(
+    v => v.host === host,
+  );
+  if (!azureIntegrationConfig) {
+    throw new Error(`Unable to locate Azure integration for the host ${host}`);
+  }
+  return azureIntegrationConfig;
+};
```

Comment:
In my opinion, these helper function can probably live inside the `@backstage/integration` package.

If a plugin wants to use the GitHub token for the host `github.com`, it should be able to call something like this without having to create a config and do the `find`.

We might also want to avoid duplications (this exact pattern is already in `backend-common`, `catalog-backend`, `catalog-import` and `scaffolder-backend` packages)

Reply:
Yep I haven't had the time yet to properly make the "unifying" class around these. Backend plugins should be given a new member of the context (e.g. in packages/backend) called `integrations`. On that one you should be able to do `integrations.byUrl(url)...` for example. The tricky bit is that the different integrations have different quirks. So to make one unifying return type of that method, is a bit finicky for anything nontrivial.